src/libAtomVM/nifs.c

```diff
@@ -2031,23 +2069,58 @@ static term nif_erlang_atom_to_binary_2(Context *ctx, int argc, term argv[])
     term atom_term = argv[0];
     VALIDATE_VALUE(atom_term, term_is_atom);
 
-    if (UNLIKELY(argv[1] != LATIN1_ATOM)) {
-        RAISE_ERROR(BADARG_ATOM);
-    }
+    term encoding = argv[1];
 
     GlobalContext *glb = ctx->global;
 
     int atom_index = term_to_atom_index(atom_term);
     size_t atom_len;
     atom_ref_t atom_ref = atom_table_get_atom_ptr_and_len(glb->atom_table, atom_index, &atom_len);
 
+    bool encode_to_latin1 = false;
+    if (encoding == LATIN1_ATOM) {
+        if (UNLIKELY(!atom_table_is_atom_ref_ascii(glb->atom_table, atom_ref))) {
```

Comment:
Avoiding an additional temporary buffer simplifies the code, I don't need to take care of allocating/deallocating/checking an additional temporary buffer, and all I need is to implement a very simple function.

> If so, maybe it could also return the string length to allocate the latin1 representation.
Alternatively, it could just compute the string length and it is ascii (7 bits) if string length is equal to bytes length.

This is an option, but it doesn't simplify a lot the code, and I compute a value that is not needed for the most common case.